src/ImageSharp/Processing/Processors/Transforms/AffineProcessor.cs

```diff
@@ -0,0 +1,304 @@
+        /// <param name="sourceMax">The maximum source bounds</param>
+        /// <param name="point">The transformed point dimension</param>
+        /// <param name="sampler">The sampler</param>
+        /// <param name="scale">The transformed image scale relative to the source</param>
+        /// <param name="weights">The collection of weights</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void CalculateWeightsDown(int min, int max, int sourceMin, int sourceMax, float point, IResampler sampler, float scale, Span<float> weights)
+        {
+            float sum = 0;
+            ref float weightsBaseRef = ref weights[0];
+
+            // Downsampling weights requires more edge sampling plus normalization of the weights
+            for (int x = 0, i = min; i <= max; i++, x++)
+            {
+                int index = i;
+                if (index < sourceMin)
+                {
+                    index = sourceMin;
+                }
+
+                if (index > sourceMax)
+                {
+                    index = sourceMax;
+                }
+
+                float weight = sampler.GetValue((index - point) / scale);
+                sum += weight;
+                Unsafe.Add(ref weightsBaseRef, x) = weight;
+            }
+
+            if (sum > 0)
+            {
+                for (int i = 0; i < weights.Length; i++)
+                {
+                    ref float wRef = ref Unsafe.Add(ref weightsBaseRef, i);
+                    wRef = wRef / sum;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Calculated the weights for the given point.
+        /// </summary>
+        /// <param name="sourceMin">The minimum source bounds</param>
+        /// <param name="sourceMax">The maximum source bounds</param>
+        /// <param name="point">The transformed point dimension</param>
+        /// <param name="sampler">The sampler</param>
+        /// <param name="weights">The collection of weights</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void CalculateWeightsScaleUp(int sourceMin, int sourceMax, float point, IResampler sampler, Span<float> weights)
+        {
+            ref float weightsBaseRef = ref weights[0];
+            for (int x = 0, i = sourceMin; i <= sourceMax; i++, x++)
+            {
+                float weight = sampler.GetValue(i - point);
+                Unsafe.Add(ref weightsBaseRef, x) = weight;
+            }
+        }
+
+        /// <summary>
+        /// Creates a new target canvas to contain the results of the matrix transform.
+        /// </summary>
+        /// <param name="sourceRectangle">The source rectangle.</param>
+        private void ResizeCanvas(Rectangle sourceRectangle)
+        {
+            this.transformMatrix = this.GetTransformMatrix();
+
+            this.targetRectangle = Matrix3x2.Invert(this.transformMatrix, out Matrix3x2 sizeMatrix)
```

Comment:
Maybe we should allow users to define their own `targetRectangle`. Some of them might need deterministic output sizes instead of automatically fitting the transformed image. This would allow them to crop or extend the result image.